graphjet-core/src/main/java/com/twitter/graphjet/algorithms/PageRank.java

```diff
@@ -0,0 +1,156 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.twitter.graphjet.algorithms;
+
+import com.google.common.util.concurrent.AtomicDouble;
+import com.twitter.graphjet.bipartite.api.EdgeIterator;
+import com.twitter.graphjet.directed.api.OutIndexedDirectedGraph;
+import it.unimi.dsi.fastutil.longs.LongArrayList;
+import it.unimi.dsi.fastutil.longs.LongIterator;
+import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
+
+/**
+ * Implementation of PageRank. This implementation initializes an array of doubles to hold PageRank values, where the
+ * node id is used as the index into the array. Therefore, the algorithm needs to know the max node id as well as the
+ * set of nodes, in the case where the node ids are not assigned sequentially.
+ */
+public class PageRank {
+  final private OutIndexedDirectedGraph graph;
+  final private LongOpenHashSet nodes;
+  final private long maxNodeId;
+  final private double dampingFactor;
+  final private int nodeCount;
+  final private int maxIterations;
+  final private double tolerance;
+
+  final private AtomicDouble normL1 = new AtomicDouble(Double.MAX_VALUE);
+  private double[] prVector = null;
+
+  /**
+   * Constructs this object for running PageRank over a directed graph.
+   *
+   * @param graph          the directed graph
+   * @param nodes          nodes in the graph
+   * @param maxNodesId     maximum node id
+   * @param dampingFactor  damping factor
+   * @param maxIterations  maximum number of iterations to run
+   * @param tolerance      L1 norm threshold for convergence
+   */
+  public PageRank(OutIndexedDirectedGraph graph, LongOpenHashSet nodes, long maxNodesId, double dampingFactor,
+                  int maxIterations, double tolerance) {
+    this.graph = graph;
+    this.nodes = nodes;
+    this.maxNodeId = maxNodesId;
+    this.dampingFactor = dampingFactor;
+    this.nodeCount = nodes.size();
+    this.maxIterations = maxIterations;
+    this.tolerance = tolerance;
+  }
+
+  private double computeL1Norm(double a[], double b[]) {
+    double ret = 0.0;
+    for (int i = 0; i < a.length; ++i) {
+      ret += Math.abs(a[i] - b[i]);
+    }
+    return ret;
+  }
+
+  private double[] iterate(double[] prevPR, double dampingAmount, LongArrayList noOuts) {
+    double nextPR[] = new double[(int) (maxNodeId + 1)]; // PageRank vector after the iteration.
+
+    // First compute how much mass is trapped at the dangling nodes.
+    AtomicDouble dangleSum = new AtomicDouble();
```

Comment:
Atomic because..? Not seeing where multiple threads will touch this. I don't think any of the algorithms use atomics since they're just using the (snapshot consistent) graphjet api to run all computations on a single thread..